SV pipeline clean up

SHuang-Broad opened the pull request:
A seemingly large change PR, but most changes are trivial.
The non-trivial part:

* a new tool `StructuralVariantionDiscoveryPipelineSpark` to run the whole process of SV discovery, by delegating works to `FindBreakpointEvidenceSpark` and `DiscoverVariantsFromContigAlignmentsSAMSpark`, both of which are refactored to accommodate the new tool;
* class `AlignmentRegion` is effectively moved into a new class `AlignedAssembly` (named quite close to the existing class `AlignedAssemblyOrExcuse` but will be moved into a different sub-package in a sequential PR).
* integration tests (local mode and on MiniClusters/hdfs) for all 5 major tools `FindBreakpointEvidenceSpark`, `DiscoverVariantsFromContigAlignmentsSAMSpark`, `StructuralVariantionDiscoveryPipelineSpark`, `AlignAssembledContigsSpark` and `DiscoverVariantsFromContigAlignmentsSGASpark`; a draw back is these integration tests do not test correctness of results  but simple tests if these tools run.
* various unit tests

The two paths involving use of Fermi-lite are tested to be running and generating compatible results. The path involves using SGA as the assembler is also running but generates significantly less variants. (see attached run logs).
[differentVersions.txt](https://github.com/broadinstitute/gatk/files/956271/differentVersions.txt)



The access levels of the various classes and methods are not optimal now because a serial PR that simply repackaging these classes (hence access levels must be changed) is expected to be generated immediately after this PR is approved.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
It strikes me that at one point we already had an RDD of AlignedAssemblyOrExcuse (I guess most recently in the code path in `FindBreakpointEvidenceSpark.handleAssemblies`) that we could just operate on if we could return it from that method (ie instead of returning a `List<AlignedAssemblyOrExcuse>` the FBE code would return an `RDD<AlignedAssemblyOrExcuse>`. Did you think about doing that at all? It seems like it would be more efficient than re-parallelizing the collection.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
I'd change the name of this method to describe what it does (ie 'convertAndFilterAlignedAssemblyOrExcuseList` or something like that).

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
Please change the names of these methods to describe what they do (`convertToAlignmentIntervals`)

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
Same, please give this method a more descriptive name

cwhelan commented on src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java:
```diff
@@ -5,14 +5,12 @@
 import htsjdk.samtools.SAMRecord;
 import htsjdk.samtools.SAMRecordCoordinateComparator;
 import org.apache.hadoop.conf.Configuration;
-import org.apache.hadoop.fs.FileStatus;
```
Just for the sake of making smaller pull requests and tracking changes I think I'd rather not change all of these imports in unrelated files.

SHuang-Broad replied:
reverted back for all

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSparkIntegrationTest.java:
I'd give this the name of the integration test

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSparkIntegrationTest.java:
Do we need to have a version with the API key? It doesn't seem to be used?

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSparkIntegrationTest.java:
Should this be group = "cluster"?

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSparkIntegrationTest.java:
Why not add the expected results to this test and compare to make sure the files are equal? Seems like an easy thing to do.

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSAMSparkIntegrationTest.java:
Again, do we need this?

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSGASparkIntegrationTest.java:
Again it seems like it would be straightforward to include the results, what do you think?

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSGASparkIntegrationTest.java:
Does this need to have the "cluster" group?

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/FindBreakpointEvidenceSparkIntegrationTest.java:
Maybe don't need api key

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqKmerSpark.java:
Nope. It was caused by the "repackaging" of classes that was taken out in this PR.

SHuang-Broad commented on scripts/sv/sga/svDiscoverSGA.sh:
```diff
@@ -40,7 +40,6 @@ export REF_2BIT
 export REF_INDEX_IMG
 
 baseDir=$(dirname -- "$0")
-${baseDir}/scanBam.sh "$@"
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqKmerSpark.java:
And I've been using IntelliJ's "Optimize Imports" tool under "Code"

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -46,20 +54,71 @@ protected void runTool(final JavaSparkContext ctx) {
+                            final ContigAligner contigAligner = new ContigAligner(indexImageFile);
+                            final List<AlignedAssembly> alignedAssembliesInThisPartition = new ArrayList<>(NUM_ASSEMBLIES_PER_PARTITION);
+                            iter.forEachRemaining(pair -> alignedAssembliesInThisPartition.add(contigAligner.alignContigs(Integer.valueOf(pair._1), pair._2)));
+                            return alignedAssembliesInThisPartition.iterator();});
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -46,20 +54,71 @@ protected void runTool(final JavaSparkContext ctx) {
+     * mapQual + {@link #MAPPED_CONTIG_ALIGNMENT_INTERVAL_STRING_REP_FIELD_SEPARATOR} + mismatch
+     */
+    @VisibleForTesting
+    public static Iterator<String> formatAlignedAssemblyAsHadoopTextFileStringList(final AlignedAssembly alignedAssembly) {
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -46,20 +54,71 @@ protected void runTool(final JavaSparkContext ctx) {
+                                        StringUtils.join(Arrays.asList(String.valueOf(alignmentInterval.startInAssembledContig) + "-" + String.valueOf(alignmentInterval.endInAssembledContig),
+                                                encodeSimpleIntervalAsString(alignmentInterval.referenceInterval),TextCigarCodec.encode(alignmentInterval.cigarAlong5to3DirectionOfContig),
+                                                (alignmentInterval.forwardStrand ? "+" : "-"), alignmentInterval.mapQual, alignmentInterval.mismatches), MAPPED_CONTIG_ALIGNMENT_INTERVAL_STRING_REP_FIELD_SEPARATOR)
+                        ).collect(Collectors.toList()), "\t");
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -46,20 +54,71 @@ protected void runTool(final JavaSparkContext ctx) {
+                .collect(Collectors.toList()).iterator();
+    }
+
+    public static String encodeSimpleIntervalAsString(final SimpleInterval simpleInterval) {
```
I wanted to, but impossible because of the HG38 contig names such as 
```
##contig=<ID=HLA-DRB1*16:02:01,length=11005,assembly=38>
```

SHuang-Broad replied:
Added documentation to explain.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -46,20 +54,71 @@ protected void runTool(final JavaSparkContext ctx) {
+    }
+
+    @VisibleForTesting
+    public static SimpleInterval decodeStringAsSimpleInterval(final String text) {
```
see above

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -46,20 +54,71 @@ protected void runTool(final JavaSparkContext ctx) {
+        }
+    }
+
+    public static final String MAPPED_CONTIG_ALIGNMENT_INTERVAL_STRING_REP_FIELD_SEPARATOR = "%";
```
I wanted the constants to be closer, physically, to their uses. But just moved.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignedAssembly.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignedAssembly.java:
Just moved `AlignedContig` to top-level, `AlignmentInterval`, if top-level, would be the same as `AlignmentRegion`.
I've discussed this with Ted and came to this compromise, which conveys the relation of "assembly is composed of contigs, whose essential information, in the discovery stage as it stands now, is the contig sequence and its (stripped down) alignment information" (A maybe unrelated issue is that we might want to reach out to other information stored in the assembly graph, but that's for later)
@tedsharpe mind chime in?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignedAssemblyOrExcuse.java:
```diff
@@ -111,41 +110,70 @@ public FermiLiteAssembly getAssembly() {
     /**
+     * Filters out failed assemblies and turn each assembly into an iterable of SAMRecords.
+     */
+    public static List<Iterable<SAMRecord>> turnIntoSAMRecordsForEachAssembly(final SAMFileHeader header, final List<AlignedAssemblyOrExcuse> alignedAssemblyOrExcuseList) {
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointComplications.java:
```diff
@@ -48,6 +42,16 @@
     private List<String> cigarStringsForDupSeqOnCtg = null;
     private boolean dupAnnotIsFromOptimization = false;
 
+    public String onErrStringRep() {
```
It used to have the `toString()` overridden, this change was to signify that this methods is to be used for human consumption when things go wrong, hence the name "on error string representation"

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
changed back.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSGASpark.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVDUSTFilteredKmerizer.java:
Added back, but IDE tells me it is not used.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
That faster route was shot down in the previous PR.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
The faster route is implemented in `directToAlignmentInterval()` below.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
I did want to do that, but resisted the temptation because this refactor PR doesn't want to mess with evidence finding stage too much.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssemblyAlignmentParser.java:
After seeking suggestion from @tedsharpe , I've replaced the `Interface` with an abstract class `AlignedContigGenerator`. The intention was really to signify an adaptation of various evidence-finding stage formats (SAM, BwaMemAilgnment, SGA-route-generated text records) to a single discovery stage format.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -25,36 +23,45 @@ public ContigAligner(final String indexImageFile) {
-     */
-    public List<AlignmentRegion> alignContigs(final String assemblyId, final ContigsCollection contigsCollection) {
-        final List<AlignmentRegion> alignedContigs = new ArrayList<>(contigsCollection.getContents().size());
+     *  @param assemblyId An identifier for the assembly or set of contigs
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSAMSpark.java:
done with extending new abstract class.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantConsensusDiscovery.java:
Updated.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVariantConsensusDiscovery.java:
removed

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
done

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSparkIntegrationTest.java:
removed

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSparkIntegrationTest.java:
done

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSparkIntegrationTest.java:
The `MiniCluster` is not the Spark cluster that I assumed it to be, it is essentially a mock up of HDFS, hence the grouping.

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSAMSparkIntegrationTest.java:
removed

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSGASparkIntegrationTest.java:
see above.

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/FindBreakpointEvidenceSparkIntegrationTest.java:
removed

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqKmerSpark.java:
Will revert back in a later commit in trying to keep this PR small.

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSparkIntegrationTest.java:
done

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSGASparkIntegrationTest.java:
done

SHuang-Broad commented:
@cwhelan , addressed comments in 4 commits, with the first 3 addressing requested changes in main and the last dealing with integration test.
Please review again. Thanks!

SHuang-Broad commented:
Also @tedsharpe would you please take a look at the changes in `FindBreakpointEvidenceSpark` and `AlignedAssemblyOrExcuse`, especially the latter since it is not a trivial change? Thanks!

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignedAssemblyOrExcuse.java:
```diff
@@ -111,41 +114,80 @@ public FermiLiteAssembly getAssembly() {
-                    }
-                }
-            }
+            filterFailedAssemblyAndConvertToSamRecords(header, alignedAssemblyOrExcuseList).forEach(contigs -> contigs.forEach(ls -> ls.forEach(writer::addAlignment)) );
```
This line is too long.  Our style guide says 100 for max line length.  But I'd be fine with the IntelliJ default of 120.

Personally, I find the functional-style breakout to be harder to read, and more obscure than the original code, but I won't object on that score.  *De gustibus non est disputandem.*  I agree that the original code -- with three nested for loops -- was too dense and complex.
What I will object to is gathering a List<List<List<SAMRecord>>> and then unfolding it all in this line.  What should be done instead, IMHO, is to start with the alignedAssemblyOrExcuseList, and then...
```
.filter( failed assemblies )
.map( to a stream of assemblies )
.flatmap( to a stream of contigs )
.flatmap( to a stream of alignments )
.map( each alignment to a SAMRecord )
.foreach( add to the writer )
```
The original code required virtually no extra working memory (no temporary collections, anyway).
Your revision requires everything to live in memory all at once before any writing is done.
The revised functional version sketched above requires minimal extra working memory.

SHuang-Broad replied:
I was actually using such style that you suggested, but the same code block is used in a different place `StructuralVariationDiscoveryPipelineSpark.InMemoryAlignmentParser.filterAndConvertToAlignedContigViaSAM()`, which was trying to go through the assemblies one by one. For the code to accommodate the needs there, it needs at least a `List<List<SAMRecord>>` where the outer level is for contigs. If everything alignment is to be flattened out to a single list, it would be difficult for that piece of code knowing which SAMRecord comes from which contig.
So I'll go with the 2-D list.

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/FindBreakpointEvidenceSpark.java:
```diff
@@ -45,129 +51,15 @@
         programGroup=StructuralVariationSparkProgramGroup.class)
 public final class FindBreakpointEvidenceSpark extends GATKSparkTool {
     private static final long serialVersionUID = 1L;
+    private final Logger localLogger = LogManager.getLogger(FindBreakpointEvidenceSpark.class);
```
Probably unnecessary for this to be an instance field -- all the methods are static, anyway.  Maybe just call getLogger in runTool?
But I'm fine with the changes to this class.

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSAMSpark.java:
```diff
@@ -23,6 +23,7 @@
 import org.broadinstitute.hellbender.exceptions.GATKException;
 import org.broadinstitute.hellbender.utils.Utils;
 import org.broadinstitute.hellbender.utils.read.GATKRead;
+import scala.Serializable;
```
Do you want `scala.Serializable` or `java.io.Serializable`?

SHuang-Broad replied:
Sorry I forgot to replace, will be removed in the final push.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/DiscoverVariantsFromContigAlignmentsSAMSpark.java:
```diff
@@ -69,23 +70,31 @@ public ReadFilter makeReadFilter() {
+            this.toolLogger = toolLogger;
+        }
 
+        @SuppressWarnings("unchecked")
```
Do you still need to suppress this warning?

SHuang-Broad replied:
Sorry I forgot to remove this in the commit, will be removed in the final push.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVVCFWriter.java:
```diff
@@ -89,10 +90,10 @@ private static void writeVariants(final PipelineOptions pipelineOptions, final S
         header.setSequenceDictionary(referenceSequenceDictionary);
         header.addMetaDataLine(VCFStandardHeaderLines.getInfoLine(VCFConstants.END_KEY));
-        GATKSVVCFHeaderLines.vcfHeaderLines.values().forEach(header::addMetaDataLine);
+//        GATKSVVCFHeaderLines.vcfHeaderLines.values().forEach(header::addMetaDataLine);
```
Why did you comment this line out?

SHuang-Broad replied:
oops, forgot to remove it.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
```diff
@@ -89,49 +87,76 @@ protected void runTool( final JavaSparkContext ctx ) {
+            this.toolLogger = toolLogger;
+        }
+
+        @SuppressWarnings("unchecked")
```
I'm not sure you need this suppression any more

SHuang-Broad replied:
removed

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSparkIntegrationTest.java:
```diff
@@ -134,8 +138,34 @@ public void testSVDiscoverPipelineRunnableMiniCluster(final StructuralVariationD
+        CloseableIterator<VariantContext> iterator;
+        final List<VariantContext> actualVcs;
+        if (onHDFS) {
+            final File tempLocalVCF = BaseTest.createTempFile("variants", "vcf"); tempLocalVCF.deleteOnExit();
```
split these statements out on to separate lines

SHuang-Broad replied:
done

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSparkIntegrationTest.java:
```diff
@@ -134,8 +138,34 @@ public void testSVDiscoverPipelineRunnableMiniCluster(final StructuralVariationD
+        final List<VariantContext> actualVcs;
+        if (onHDFS) {
+            final File tempLocalVCF = BaseTest.createTempFile("variants", "vcf"); tempLocalVCF.deleteOnExit();
+            new Path(generatedVCFPath).getFileSystem(new Configuration()).copyToLocalFile(new Path(generatedVCFPath), new Path(tempLocalVCF.getAbsolutePath()));
```
I think you can just use `BucketUtils.copyFile` and not have to worry about getting the HDFS `FileSystem` object here.

SHuang-Broad replied:
done

SHuang-Broad commented:
@tedsharpe please see if you are OK with the latest commit by replacing the 3-D list with a stream.
@cwhelan I'll address your comments in a separate commit.
Thank you both!
